### PR TITLE
isStandard compliance [for pubKeyHash / scriptHash outputs]

### PR DESCRIPTION
**Pull request by dcousens**
As pointed out [here](https://github.com/bitcoinjs/bitcoinjs-lib/issues/480),  this is an interesting issue.

`script.decompile` will allow for non-minimal `OP_PUSHDATA`s (that is,  you can specify a larger `OP_PUSHDATAx` than needed),  meaning extra data can be added into the script.

In some cases (as pointed out in https://github.com/bitcoinjs/bitcoinjs-lib/issues/480),  this can be confusing because [bitcoin/bitcoin](https://github.com/bitcoin/bitcoin/blob/2b0567b00291205731cbb910fb3cf33359a046ee/src/script/standard.cpp#L61) doesn't consider these transactions standard.

So far, bitcoinjs-lib hasn't cared either way, and has allowed the construction of non-standard transactions.

Perhaps an `IsStandard` flag that is enabled by default on these scripts?
Thoughts?

**Or,  better yet,  perhaps our `is*` functions are assumed to be adhering to `IsStandard`, so naturally all we have to do is amend the functions as per this PR?**

@fanatid curious on your opinion here.

### CODE REVIEW

**Comment by dcousens** [comment]
But then again,  that would mean `isScriptHashInput` is inconsistent, because right now it doesn't care about standard rules and checks generically.

**Comment by fanatid** [comment]
Not sure about this, but I think better way will be add `isStandard` (or `strict`) argument to `is*` functions and to `docompile` (when `isStandard` is `true` `decompile` following to minimal OP_PUSHDATA compliance).
Example for `isScriptHashOutput`:
```js
function isScriptHashOutput (script, isStandard) {
  var chunks = decompile(script, isStandard)
  return chunks.length === 3 &&
    chunks[0] === OPS.OP_HASH160 &&
    Buffer.isBuffer(chunks[1]) &&
    chunks[1].length === 20 &&
    chunks[2] === OPS.OP_EQUAL
}
```

In this case it's allowed to check scripts in strict mode and have backward compatibility with old code (`isStandard` should be `false` by default).

**Comment by dcousens** [comment]
Thought about this for a few hours,  and,  it makes sense to make the `is*` functions mostly compliant with `IsStandard` rules simply because the entire concept of a `pay-to-pubkeyhash` and `pay-to-scripthash` is about following the community set standards for those functions.

**Comment by dcousens** [comment]
**TL;DR**:
Considering the `is*` functions are used to match scripts against community set script templates,  adhering to the bitcoin core `IsStandard` rules makes sense IMHO.

To match otherwise, you should probably use a script interpreter or write your own function.

**Comment by fanatid** [comment]
I think that `isCanonicalPubKey` also must need changes... I write some closely code in [script2addresses](https://github.com/fanatid/script2addresses/blob/master/src/publickey.js)

**Comment by dcousens** [comment]
@fanatid what do you mean?  The code for `isCanonicalPubKey` mirrors the code from https://github.com/bitcoin/bitcoin/blob/master/src/script/interpreter.cpp#L67-L87

**Comment by fanatid** [comment]
hm, may be I am wrong.
but in different places you can see different rules:
[size is 33 <= x <= 65 (first byte doesn't important)](https://github.com/bitcoin/bitcoin/blob/0fbfc5106cd9866325b4a1ab3b9db8e91e54f070/src/script/standard.cpp#L129)
[size is 33 (first byte 2 or 3) and 65 (first byte is 4, 6 or 7)](https://github.com/bitcoin/bitcoin/blob/0fbfc5106cd9866325b4a1ab3b9db8e91e54f070/src/pubkey.h#L48) ([called from TransactionSignatureChecker::CheckSig](https://github.com/bitcoin/bitcoin/blob/0fbfc5106cd9866325b4a1ab3b9db8e91e54f070/src/script/interpreter.cpp#L1111))
[and finally in secp256k1 with additional checks](https://github.com/bitcoin/bitcoin/blob/0fbfc5106cd9866325b4a1ab3b9db8e91e54f070/src/secp256k1/src/eckey_impl.h#L17)
I don't think, that is right to take example from `interpreter.cpp` for `isCanonicalPubKey`.

**Comment by dcousens** [comment]
@fanatid remember this isn't a script interpreter, but just a parser.
If you want to verify the public key is a valid ECDSA public key, then you will need to do that yourself (and naturally handle each error case that could consist of).

**Comment by dcousens** [comment]
Going to merge this.
I think the desired result is that the script templates we are checking against are [and were] intended for verifying your own scripts against the `isStandard` equivalents.
In this sense, a `isStandard` flag isn't necessary.

If you want to be *more* liberal in what you check/verify against, it is perfectly simple to write your own.
Therefore, I'm going to merge this on the assumption that it is a patch for previously broken behaviour, not a breaking change.

**Comment by dcousens** [comment]
@fanatid `SCRIPT_VERIFY_STRICTENC` is by default on in bitcoin core, which is what we are matching in terms of these script templates.
Therefore, I think comparing against [interpreter.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/script/interpreter.cpp#L67-L87) is fine.
